src/node_i18n.cc

```diff
@@ -62,6 +72,306 @@ bool flag_icu_data_dir = false;
 
 namespace i18n {
 
+#define STRINGLEN(isolate, buf, len)                                          \
+  String::NewFromUtf8(isolate, buf, v8::NewStringType::kNormal, len)          \
+      .ToLocalChecked()
+
+#define STRING(isolate, buf)                                                  \
+  String::NewFromUtf8(isolate, buf, v8::NewStringType::kNormal, len)          \
+      .ToLocalChecked()
+
+const size_t kStorageSize = 1024;
+
+void ThrowICUError(Environment* env, UErrorCode status, const char* msg) {
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  if (msg == nullptr || msg[0] == '\0')
+    msg = "Unspecified ICU Exception";
+  char buf[kStorageSize];
+  int len = snprintf(buf, sizeof(buf), "%s [%s]", msg, u_errorName(status));
+  Local<Value> e = Exception::Error(STRINGLEN(isolate, buf, len));
+  if (e.IsEmpty())
+    e = Local<Value>();
+  Local<Object> obj = e.As<Object>();
+  obj->Set(env->context(),
+           env->errno_string(),
+           Integer::New(isolate, status)).FromJust();
+  obj->Set(env->context(),
+           env->code_string(),
+           STRING(isolate, u_errorName(status))).FromJust();
+  env->isolate()->ThrowException(e);
+}
+
+// TODO(jasnell): This could potentially become a member of MaybeStackBuffer
+// at some point in the future. Care would need to be taken with the
+// MaybeStackBuffer<UChar> variant below.
+MaybeLocal<Object> AsBuffer(Isolate* isolate,
+                            MaybeStackBuffer<char>* buf,
+                            size_t len) {
+  if (buf->IsAllocated()) {
+    MaybeLocal<Object> ret = Buffer::New(isolate, buf->out(), len);
+    if (!ret.IsEmpty()) buf->Release();
+    return ret;
+  }
+  return Buffer::Copy(isolate, buf->out(), len);
+}
+
+MaybeLocal<Object> AsBuffer(Isolate* isolate,
+                            MaybeStackBuffer<UChar>* buf,
+                            size_t len) {
+  char* dst = reinterpret_cast<char*>(buf->out());
+  if (IsBigEndian())
+    SwapBytes16(dst, len * sizeof(UChar));
+  if (buf->IsAllocated()) {
+    MaybeLocal<Object> ret = Buffer::New(isolate, dst, len);
+    if (!ret.IsEmpty()) buf->Release();
+    return ret;
+  }
+  return Buffer::Copy(isolate, dst, len);
+}
+
+struct Converter {
+  Converter(Environment* env, const char* name)
+      : conv(nullptr) {
+    UErrorCode status = U_ZERO_ERROR;
+    conv = ucnv_open(name, &status);
+    CHECK(U_SUCCESS(status));
+  }
+
+  ~Converter() {
+    if (conv != nullptr)
```

Comment:
Isn't this logically always true because of the CHECK in the constructor?  If yes, I'd either remove this or change it to a CHECK.